components/properties/src/bidi_data.rs

```diff
@@ -0,0 +1,137 @@
+/// This module provides an efficient storage of data serving the following
+/// properties:
+/// - `Bidi_Paired_Bracket`
+/// - `Bidi_Paired_Bracket_Type`
+/// - `Bidi_Mirrored`
+/// - `Bidi_Mirroring_Glyph`
+///
+
+
+/// 20..0  Code point return value for Bidi_Mirroring_Glyph value
+/// 21..21 Boolean for Bidi_Mirrored
+/// 23..22 Enum value for Bidi_Paired_Bracket_Type
+#[doc(hidden)] // needed for datagen but not intended for users
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct MirroredPairedBracketData {
+    mirroring_glyph: char,
+    is_mirrored: bool,
+    paired_bracket_type: BidiPairedBracketType,
+}
+
+#[doc(hidden)] // needed for datagen but not intended for users
+#[derive(Copy, Clone, Hash, PartialEq, Eq, Debug)]
+#[repr(packed)]
+pub struct MirroredPairedBracketDataULE([u8; 3]);
+
+// Safety (based on the safety checklist on the ULE trait):
+//  1. MirroredPairedBracketDataULE does not include any uninitialized or padding bytes
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  2. MirroredPairedBracketDataULE is aligned to 1 byte.
+//     (achieved by `#[repr(transparent)]` on a type that satisfies this invariant)
+//  3. The impl of validate_byte_slice() returns an error if any byte is not valid.
+//  4. The impl of validate_byte_slice() returns an error if there are extra bytes.
+//  5. The other ULE methods use the default impl.
+//  6. CharULE byte equality is semantic equality
+unsafe impl ULE for MirroredPairedBracketDataULE {
+    #[inline]
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), ZeroVecError> {
+        if bytes.len() % 3 != 0 {
+            return Err(ZeroVecError::InvalidLength {
+                ty: "MirroredPairedBracketDataULE",
+                len: bytes.len(),
+            });
+        }
+        // Validate the bytes
+        for byte_triple in bytes.chunks_exact(3) {
+            // Bidi_Mirroring_Glyph validation
+            #[allow(clippy::indexing_slicing)]
+            // Won't panic because the chunks are always 3 bytes long
+            let byte0 = byte_triple[0];
+            let byte1 = byte_triple[1];
+            let byte2 = byte_triple[2];
+            let mut mirroring_glyph_code_point: u32 = (byte2 | 0x1F) as u32;
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte1 as u32);
+            mirroring_glyph_code_point = (mirroring_glyph_code_point << 8) | (byte0 as u32);
+            let _mirroring_glyph = char::from_u32(mirroring_glyph_code_point)
+                .ok_or(ZeroVecError::VarZeroVecFormatError)?;
+
+            // skip validating the Bidi_Mirrored boolean since it is always valid
+
+            // skip validating Bidi_Paired_Bracket_Type because enum is open
+            //
+            // TODO: is the above statement correct and okay?
+            // TODO: or do we have a way to validate the enum discriminant against the enum for
+            //       newtype idiom enums?
```

Comment:
issue: we must validate that all _unused_ bytes are 0 (or some other fixed value)